packages/backend/server/src/base/metrics/opentelemetry.ts

```diff
@@ -81,39 +81,58 @@ export abstract class OpentelemetryFactory {
   }
 }
 
-export class LocalOpentelemetryFactory
-  extends OpentelemetryFactory
+@Injectable()
+export class OpentelemetryFactory
+  extends BaseOpentelemetryFactory
   implements OnModuleDestroy
 {
-  private readonly metricsExporter = new PrometheusExporter({
-    metricProducers: this.getMetricsProducers(),
-  });
+  private readonly logger = new Logger(OpentelemetryFactory.name);
+  #sdk: NodeSDK | null = null;
+
+  constructor(private readonly config: Config) {
+    super();
+  }
+
+  @OnEvent('config.init')
+  async init(event: Events['config.init']) {
+    if (event.config.metrics.enabled) {
+      await this.setup();
+      registerCustomMetrics();
+    }
+  }
+
+  @OnEvent('config.changed')
+  async onConfigChanged(event: Events['config.changed']) {
+    if ('metrics' in event.updates) {
+      await this.setup();
+    }
+  }
 
   async onModuleDestroy() {
-    await this.metricsExporter.shutdown();
+    await this.#sdk?.shutdown();
   }
 
   override getMetricReader(): MetricReader {
-    return this.metricsExporter;
+    return new PrometheusExporter({
+      metricProducers: this.getMetricsProducers(),
+    });
   }
 
   override getSpanExporter(): SpanExporter {
     return new ZipkinExporter();
   }
-}
-
-function getMeterProvider() {
-  return metrics.getMeterProvider();
-}
 
-export function registerCustomMetrics() {
-  const hostMetricsMonitoring = new HostMetrics({
-    name: 'instance-host-metrics',
-    meterProvider: getMeterProvider() as MeterProvider,
-  });
-  hostMetricsMonitoring.start();
-}
-
-export function getMeter(name = 'business') {
-  return getMeterProvider().getMeter(name);
+  private async setup() {
+    if (this.config.metrics.enabled) {
+      if (!this.#sdk) {
+        this.#sdk = this.create();
+      }
+      this.#sdk.start();
+      this.logger.log('OpenTelemetry SDK started');
+    } else {
+      await this.#sdk?.shutdown();
+      this.#sdk = null;
```

Comment:
Will the sdk become null after the `config.changed` event is triggered?

Reply:
what's the point?

Reply:
it's in `enabled` branch